consensus/notify/src/notification.rs

```diff
@@ -0,0 +1,173 @@
+
+    #[display(fmt = "UtxosChanged notification")]
+    UtxosChanged(UtxosChangedNotification),
+
+    #[display(fmt = "SinkBlueScoreChanged notification: virtual selected parent blue score {}", "_0.sink_blue_score")]
+    SinkBlueScoreChanged(SinkBlueScoreChangedNotification),
+
+    #[display(fmt = "VirtualDaaScoreChanged notification: virtual DAA score {}", "_0.virtual_daa_score")]
+    VirtualDaaScoreChanged(VirtualDaaScoreChangedNotification),
+
+    #[display(fmt = "PruningPointUtxoSetOverride notification")]
+    PruningPointUtxoSetOverride(PruningPointUtxoSetOverrideNotification),
+
+    #[display(fmt = "NewBlockTemplate notification")]
+    NewBlockTemplate(NewBlockTemplateNotification),
+}
+}
+
+impl NotificationTrait for Notification {
+    fn apply_overall_subscription(&self, subscription: &OverallSubscription) -> Option<Self> {
+        match subscription.active() {
+            true => Some(self.clone()),
+            false => None,
+        }
+    }
+
+    fn apply_virtual_chain_changed_subscription(&self, subscription: &VirtualChainChangedSubscription) -> Option<Self> {
+        match subscription.active() {
+            true => {
+                if let Notification::VirtualChainChanged(ref payload) = self {
+                    if !subscription.include_accepted_transaction_ids() && !payload.accepted_transaction_ids.is_empty() {
+                        return Some(Notification::VirtualChainChanged(VirtualChainChangedNotification {
+                            removed_chain_block_hashes: payload.removed_chain_block_hashes.clone(),
+                            added_chain_block_hashes: payload.added_chain_block_hashes.clone(),
+                            accepted_transaction_ids: Arc::new(vec![]),
+                        }));
+                    }
+                }
+                Some(self.clone())
+            }
+            false => None,
+        }
+    }
+
+    fn apply_utxos_changed_subscription(&self, _subscription: &UtxosChangedSubscription) -> Option<Self> {
+        // No effort is made here to apply the subscription addresses.
+        // This will be achieved farther along the notification backbone.
+        Some(self.clone())
+    }
+
+    fn event_type(&self) -> EventType {
+        self.into()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct BlockAddedNotification {
+    pub block: Arc<Block>,
+}
+
+impl BlockAddedNotification {
+    pub fn new(block: Arc<Block>) -> Self {
+        Self { block }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct VirtualChainChangedNotification {
+    pub added_chain_block_hashes: Arc<Vec<Hash>>,
+    pub removed_chain_block_hashes: Arc<Vec<Hash>>,
+    pub accepted_transaction_ids: Arc<Vec<TransactionId>>,
+}
+
+impl VirtualChainChangedNotification {
+    pub fn new(
+        added_chain_block_hashes: Arc<Vec<Hash>>,
+        removed_chain_block_hashes: Arc<Vec<Hash>>,
+        accepted_transaction_ids: Arc<Vec<TransactionId>>,
```

Comment:
Technically this should be an array corresponding to `added_chain_block_hashes` with an `accepted_ids` vector per added chain block. Or better of, define an `AcceptanceData` struct with a field for `accepted_ids` and use it. However note that this data should be sent only if there is a listener for it (as it is heavy).

In golang we have this data pulled from consensus only if there's a listener. 

It seems like here we can check the subscription, but I emphasize that we should have a way to check the subscription and only then read this data from stores

Reply:
Partially done
`AcceptanceData` is moved to `consensus-core` and used in the notification